Add early return to WebKitCocoa API tests that fail in site isolation

colelao opened the pull request:
#### 6b6abab942f3518cf646cf837f32ed33431d6908
<pre>
Add early return to WebKitCocoa API tests that fail in site isolation
<a href="https://rdar.apple.com/161912764">rdar://161912764</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300128">https://bugs.webkit.org/show_bug.cgi?id=300128</a>

Reviewed by Sihui Liu.

This patch adds a function isSiteIsolationEnabledForSkippingTests that
is used in API tests. Tests that are known and tracked site isolation
bugs should be skipped in site isolation. This patch tracks down and skips
SOME of them.

Tests: Tools/TestWebKitAPI/SourcesCocoa.txt
       Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj
       Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.h
       Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, WillGoToBackForwardListItem)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, Back)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.h: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.mm: Added.
(isSiteIsolationEnabled):

Canonical link: <a href="https://commits.webkit.org/301145@main">https://commits.webkit.org/301145@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14261a1badcb5afc1fdb6de3bc35894b39e08875

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44605 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131786 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76857 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126813 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45298 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53172 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95084 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127890 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111735 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75632 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35077 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29888 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75264 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105904 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30119 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134457 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51766 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39564 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103573 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52187 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107948 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103339 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26341 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48678 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26970 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48785 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51645 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51027 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54382 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.mm:
```diff
@@ -23,20 +23,20 @@
-namespace WebCore {
-
-AcceleratedEffectTransformOrigin blend(const AcceleratedEffectTransformOrigin& from, const AcceleratedEffectTransformOrigin& to, const BlendingContext& context)
+bool siteIsolationEnabled(WKWebView* webView)
```
Boolean values are generally prefixed with is/did.
```suggestion
bool isSiteIsolationEnabled(WKWebView* webView)
```

pxlcoder replied:
`*` next to variable name for Objective-C types.
```suggestion
bool siteIsolationEnabled(WKWebView *webView)
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.mm:
```diff
@@ -23,20 +23,20 @@
-    return { .value = WebCore::blend(from.value, to.value, context) };
-}
+    auto configuration = webView.configuration;
+    auto preferences = configuration.preferences;
```
`RetainPtr` rather than `auto` per smart pointer guidelines.

pxlcoder commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -729,9 +729,9 @@
-		A176D24D2D6D2A4E00C63915 /* WebContentExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2492D6D295400C63915 /* WebContentExtension.appex */; platformFilter = ios; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		A176D24B2D6D2A4E00C63915 /* GPUExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2472D6D295400C63915 /* GPUExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		A176D24C2D6D2A4E00C63915 /* NetworkingExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2482D6D295400C63915 /* NetworkingExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		A176D24D2D6D2A4E00C63915 /* WebContentExtension.appex in Embed Extensions */ = {isa = PBXBuildFile; fileRef = A176D2492D6D295400C63915 /* WebContentExtension.appex */; platformFilters = (ios, ); settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
```
Please undo.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -24,10 +24,12 @@
  */
 
 #import "config.h"
+#include "gtest/gtest.h"
```
Why is this necessary? `#include` also looks out of place.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
```diff
@@ -24,12 +24,14 @@
  */
 
 #import "config.h"
+#include "gtest/gtest.h"
```
Same question.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -722,6 +723,11 @@ - (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigat
     auto webView = adoptNS([[WKWebView alloc] init]);
+    // FIXME: this test fails in site isolation. Remove the following line once addressed.
+    // In site isolation, page cache support is missing. The test should be fixed when we add support for page cache
+    if (isSiteIsolationEnabledForSkippingTest(webView.get()))
```
Let's add radar number / bug number to the FIXME comment: like 
`FIXME: Page cache is currently disabled under site isolation; see rdar://161762363.`

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -909,6 +910,11 @@ static void runBasicTest(SchemeHandlerShouldBeAsync schemeHandlerShouldBeAsync)
     [[webViewConfiguration userContentController] addScriptMessageHandler:messageHandler.get() name:@"pson"];
 
     auto webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:webViewConfiguration.get()]);
+    // FIXME: this test fails in site isolation. Remove the following line once addressed.
```
Ditto, let's add radar / bug number to clarify when we expect to unskip.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -9550,6 +9556,10 @@ HTTPServer server({
     });
 
     auto webView = adoptNS([[WKWebView alloc] initWithFrame:CGRectMake(0, 0, 800, 600)]);
+    // FIXME: this test fails in site isolation. Remove the following line once addressed.
```
Ditto.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.mm:
```diff
@@ -0,0 +1,50 @@
+ * enabled. Any test that uses this function does (or at some point did) represent
+ * an active bug in site isolation, and should be addressed.
+ */
+bool isSiteIsolationEnabledForSkippingTest(WKWebView *webView)
```
`isSiteIsolationEnabled` is probably enough (as the caller gets to decide what to do)

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:
```diff
@@ -2091,6 +2092,11 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
     auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 800, 600) configuration:configuration.get()]);
+    // FIXME: this test fails in site isolation. Remove the following line once addressed.
+    // In site isolation, results[1] == "". We don't expect document to have different referrers when site isolation is on
+    if (isSiteIsolationEnabledForSkippingTest(webView.get()))
```
Enabling page cache under SI is feature-level work so we expect it will take a while before we can run the tests, but this looks like a bug fix, so for this one I suggest fixing instead of skipping.

colelao committed:
Add early return to WebKitCocoa API tests that fail in site isolation
rdar://161912764
https://bugs.webkit.org/show_bug.cgi?id=300128

Reviewed by Sihui Liu.

This patch adds a function isSiteIsolationEnabledForSkippingTests that
is used in API tests. Tests that are known and tracked site isolation
bugs should be skipped in site isolation. This patch tracks down and skips
SOME of them.

Tests: Tools/TestWebKitAPI/SourcesCocoa.txt
       Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj
       Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.h
       Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, WillGoToBackForwardListItem)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, Back)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.h: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolationUtilities.mm: Added.
(isSiteIsolationEnabled):

Canonical link: https://commits.webkit.org/301145@main
